### PR TITLE
UCS/MEMTRACK: added public API for memtrack:

### PR DESCRIPTION
**Pull request by hoopoepg**
- routines ucs_memtrack_allocated and ucs_memtrack_releasing moved
  to public header
- added internal memtrack header
- memtrack configure option is removed, memtrack
  now is always built-in

### CODE REVIEW

**Comment by hoopoepg** [comment]
bot:pipe:retest

File: src/ucs/datastruct/ptr_array.c [hunk 1]
```diff
@@ -107,18 +107,17 @@ static void UCS_F_MAYBE_UNUSED ucs_ptr_array_dump(ucs_ptr_array_t *ptr_array)
...
+    ptr_array->name     = NULL;
 }
 
 void ucs_ptr_array_init(ucs_ptr_array_t *ptr_array, const char *name)
 {
     ucs_ptr_array_clear(ptr_array);
-#ifdef ENABLE_MEMTRACK
-    ucs_snprintf_zero(ptr_array->name, sizeof(ptr_array->name), "%s", name);
-#endif
+    ptr_array->name = ucs_strdup(name, "ptr_array name");
```
Comment (yosefe) [inline src/ucs/datastruct/ptr_array.c:120]: maybe ptr_array->name = name

File: src/ucs/datastruct/ptr_array.c [hunk 2]
```diff
@@ -136,17 +135,18 @@ void ucs_ptr_array_cleanup(ucs_ptr_array_t *ptr_array)
...
 
     ucs_free(ptr_array->start);
+    ucs_free((void*)ptr_array->name);
     ucs_ptr_array_clear(ptr_array);
 }
 
-static void ucs_ptr_array_grow(ucs_ptr_array_t *ptr_array, unsigned new_size
-                               UCS_MEMTRACK_ARG)
+static void ucs_ptr_array_grow(ucs_ptr_array_t *ptr_array, unsigned new_size,
+                               const char* alloc_name)
```
Comment (yosefe) [inline src/ucs/datastruct/ptr_array.c:143]: maybe not pass alloc_name. take it from ptr_array->name

File: src/ucs/debug/memtrack.c [hunk 3]
```diff
@@ -172,7 +170,7 @@ static void ucs_memtrack_generate_report()
     }
 }
 
-void ucs_memtrack_allocated(void *ptr, size_t size, const char *name)
+static void ucs_memtrack_do_allocated(void *ptr, size_t size, const char *name)
```
Comment (yosefe) [inline src/ucs/debug/memtrack.c:173]: UCS_F_NOINLINE

File: src/ucs/debug/memtrack.c [hunk 4]
```diff
@@ -224,16 +218,23 @@ void ucs_memtrack_allocated(void *ptr, size_t size, const char *name)
...
 }
 
-void ucs_memtrack_releasing(void* ptr)
-{
-    ucs_memtrack_entry_t *entry;
-    khiter_t iter;
-    size_t size;
 
+static UCS_F_ALWAYS_INLINE void
+ucs_memtrack_allocated_int(void *ptr, size_t size, const char *name)
```
Comment (yosefe) [inline src/ucs/debug/memtrack.c:223]: int->internal

File: src/ucs/debug/memtrack.c [hunk 5]
```diff
@@ -255,25 +256,34 @@ void ucs_memtrack_releasing(void* ptr)
     pthread_mutex_unlock(&ucs_memtrack_context.lock);
 }
 
+static UCS_F_ALWAYS_INLINE void ucs_memtrack_releasing_int(void *ptr)
+{
+    if ((ptr == NULL) || !ucs_memtrack_is_enabled()) {
```
Comment (yosefe) [inline src/ucs/debug/memtrack.c:261]: maybe check for ptr==NULL inside do_{allocated|releasing} functions ?

File: src/ucs/debug/memtrack_int.h [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+ * Track custom allocation. Need to be called after custom allocation returns.
+ */
+void ucs_memtrack_allocated(void *ptr, size_t size, const char *name);
+
+
+/**
+ * Track release of custom allocation. Need to be called before actually
+ * releasing the memory.
+ */
+void ucs_memtrack_releasing(void *ptr);
```
Comment (yosefe) [inline src/ucs/debug/memtrack_int.h:87]: why needed here?

File: src/ucs/debug/memtrack_int.h [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+int ucs_munmap(void *addr, size_t length);
+char *ucs_strdup(const char *src, const char *name);
+char *ucs_strndup(const char *src, size_t n, const char *name);
+
+END_C_DECLS
+
+/*
+ * The functions below have no native implementation, they apply to both cases.
+ */
+int ucs_posix_memalign_realloc(void **ptr, size_t boundary, size_t size,
```
Comment (yosefe) [inline src/ucs/debug/memtrack_int.h:111]: move it before END_C_DECLS?

File: src/ucs/datastruct/ptr_array.c (hunk 1 above)
Comment (hoopoepg) [reply src/ucs/datastruct/ptr_array.c]: > maybe ptr_array->name = name

ok, done

File: src/ucs/datastruct/ptr_array.c (hunk 2 above)
Comment (hoopoepg) [reply src/ucs/datastruct/ptr_array.c]: > maybe not pass alloc_name. take it from ptr_array->name

done

File: src/ucs/debug/memtrack.c (hunk 4 above)
Comment (hoopoepg) [reply src/ucs/debug/memtrack.c]: > int->internal

renamed

File: src/ucs/debug/memtrack.c (hunk 5 above)
Comment (hoopoepg) [reply src/ucs/debug/memtrack.c]: > maybe check for ptr==NULL inside do_{allocated|releasing} functions ?

done

File: src/ucs/debug/memtrack_int.h (hunk 6 above)
Comment (hoopoepg) [reply src/ucs/debug/memtrack_int.h]: > why needed here?

copy/paste issue

File: src/ucs/debug/memtrack_int.h (hunk 7 above)
Comment (hoopoepg) [reply src/ucs/debug/memtrack_int.h:96]: > move it before END_C_DECLS?

it was below macro in original code, I expected it has "specific" implementation...
fixed

**Review by yosefe — APPROVED** [review]

**Comment by hoopoepg** [comment]
@yosefe I fixed Coverity issue
ok to squash?

**Review by yosefe — APPROVED** [review]

**Comment by hoopoepg** [comment]
@yosefe ok to merge?

**Pull request merged**